### PR TITLE
RACSignal+Operations: rewrite -switchToLatest without -takeUntil:.

### PR DESCRIPTION
**Pull request by StatusReport**
The current implementation of `-switchToLatest` uses the built in
`-flattenMap:` and `-takeUntil:` operators. Besides performance
overhead, there are inherent issues in the current implementation:

1. The current active inner signal is not being disposed before a new
   inner signal is activated and subscribed to (see test for a
   reproduction).

2. A race condition is possible if the inner and outer signals are
   running on different schedulers. See this for more info:
   https://github.com/ReactiveCocoa/ReactiveCocoa/issues/2781

Note that while the new implementation solves (1), (2) cannot be
hermetically solved until `RACPassthroughSubscriber` will enforce
correct disposal of its inner subscription.

The performance of the new operator was tested both on Simulator and
iPhone X with the following code:

```objc
RACSubject *subject = [RACSubject subject];
RACSignal *stl = [subject switchToLatest];

[stl subscribeNext:^(id _Nullable) {
}];

for (int i = 0; i < 1000; ++i) {
  auto inner = [RACSubject subject];
  [subject sendNext:inner];

  for (int j = 0; j < 1000; ++j) {
    [inner sendNext:@(j)];
  }
}
```

And the measurements:

Simulator: 10028.4 ms (old) vs. 6670.07 ms (new) - 33% improvement.
iPhone X: 1407.5 ms (old vs. 903.832 ms (new) - 37% improvement.

### CODE REVIEW

**Comment by StatusReport** [comment]
@barakwei + @Shachlan, please review as well as this is one nasty and critical piece of code.

BTW, notice the performance of simulator vs. device here. I have no explanation to why it's x10 faster.

File: ReactiveObjC/RACSignal+Operations.m [hunk 1]
```diff
@@ -846,22 +846,57 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
...
 
-    RACDisposable *subscriptionDisposable = [[connection.signal
-      flattenMap:^(RACSignal *x) {
-        NSCAssert(x == nil || [x isKindOfClass:RACSignal.class], @"-switchToLatest requires that the source signal (%@) send signals. Instead we got: %@", self, x);
+    RACSerialDisposable *innerDisposable = [[RACSerialDisposable alloc] init];
+    NSLock *lock = [[NSLock alloc] init];
+
+    RACDisposable *outerDisposable = [self subscribeNext:^(RACSignal *signal) {
+      NSCAssert(signal == nil || [signal isKindOfClass:RACSignal.class], @"-switchToLatest expects signals as values, instead we got: %@", signal);
+      signal = signal ?: [RACSignal empty];
```
Comment (Shachlan) [inline ReactiveObjC/RACSignal+Operations.m]: minor: This can be moved to before the assert, in order to save the nil check in the assert.

File: ReactiveObjC/RACSignal+Operations.m [hunk 2]
```diff
@@ -846,22 +846,57 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
...
+          if (outerCompleted) {
+            [subscriber sendCompleted];
+          } else {
+            innerCompleted = YES;
+          }
+        }
+        [lock unlock];
+      }];
+
+      [innerDisposable swapInDisposable:disposable];
```
Comment (Shachlan) [inline ReactiveObjC/RACSignal+Operations.m]: minor: I'm not sure that there's a concrete benefit to use a serial disposable over just a __block variable, since you're already disposing of the internal disposable beforehand - you're just incurring extra cost by using the atomic protections.

File: ReactiveObjCTests/RACSignalSpec.m [hunk 3]
```diff
@@ -1984,6 +1984,20 @@ + (void)configure:(Configuration *)configuration {
     [subject sendCompleted];
     expect(@(completed)).to(beTruthy());
   });
+	
+  qck_it(@"should dispose previous inner signal before subscribing to new inner signal", ^{
+    RACSubject *otherSignal = [RACSubject subject];
+    RACSignal *signal = [RACSignal createSignal:^RACDisposable *(id<RACSubscriber> subscriber) {
+      [otherSignal sendNext:@"foo"];
```
Comment (Shachlan) [inline ReactiveObjCTests/RACSignalSpec.m]: minor: consider adding a test where otherSignal sends from another thread, in order to test for thread safety.

File: ReactiveObjC/RACSignal+Operations.m [hunk 4]
```diff
@@ -846,22 +846,57 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
...
+      NSUInteger signalIndex = innerSignalIndex;
+      [lock unlock];
+
+      RACDisposable *disposable = [signal subscribeNext:^(id x) {
+        [subscriber sendNext:x];
+      } error:^(NSError *error) {
+        [subscriber sendError:error];
+      } completed:^{
+        [lock lock];
+        if (innerSignalIndex == signalIndex) {
```
Comment (barakyoresh) [inline ReactiveObjC/RACSignal+Operations.m]: All of this index shenanigans is only to overcome the potential sending of values after disposal?

Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: Without this, it's possible for the previous inner signal to complete and then if the outer signal completes, the signal will complete. This is although there's a new inner signal which is still active.

I know it's a bit clunky. Let me know if you have a more elegant solution here.

Comment (barakyoresh) [reply ReactiveObjC/RACSignal+Operations.m]: Yes but you update the index after you dispose the previous signal. So my question is: can it still potentially send values after disposal?

**Comment by StatusReport** [comment]
@barakwei I agree, that was one thing I was going to change, although I need to think a bit more whether it opens up the possibility of double/unordered completions.

**Comment by barakwei** [comment]
@StatusReport you might want to use the boolean flag to keep our new signals as well

File: ReactiveObjC/RACSignal+Operations.m [hunk 5]
```diff
@@ -846,22 +846,57 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
...
-        return [x takeUntil:[connection.signal concat:[RACSignal never]]];
-      }]
-      subscribe:subscriber];
+      [innerDisposable.disposable dispose];
+
+      [lock lock];
+      innerCompleted = NO;
+      ++innerSignalIndex;
+      NSUInteger signalIndex = innerSignalIndex;
+      [lock unlock];
```
Comment (barakwei) [inline ReactiveObjC/RACSignal+Operations.m]: Might need to move this until after the `subscribeNext`.

Comment (Shachlan) [reply ReactiveObjC/RACSignal+Operations.m]: but then values that are sent synchronously will be sent within the lock. that sounds quite dangerous.

Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: I don't think that it's required. See my comment below on the order of things.

File: ReactiveObjC/RACSignal+Operations.m (hunk 4 above)
Comment (barakyoresh) [reply ReactiveObjC/RACSignal+Operations.m]: I see that's what I figured.

Wouldn't it be more straightforward to store the pointer of the current active inner signal?

Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: Glad you asked! This may not work since no one guarantees that pointers will not be reused in another allocation.

We may be able to hold the signal weakly and then compare, though. Let me know if you find any loopholes here.

Comment (barakyoresh) [reply ReactiveObjC/RACSignal+Operations.m]: Can you point me in the direction of a trustworthy resource regarding pointer reusability? Couldn't find anything. 

If @Shachlan is stressed about this thing I can let this counter blasphemy slide.

Comment (Shachlan) [reply ReactiveObjC/RACSignal+Operations.m]: Yes, this delays a version. I'd like it to be merged quickly.

Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: I don't know any source that explains it, but it's very simple - the signal you hold the pointer of can get deallocated (since it's not being held by ARC) and then the pointer can get reassigned to a different object. By comparing pointers you don't really know that one of the pointers got replaced with a completely different object.

Using `__weak` should solve this as the original pointer will get set to `nil`.

File: ReactiveObjC/RACSignal+Operations.m (hunk 1 above)
Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: Done.

File: ReactiveObjC/RACSignal+Operations.m (hunk 2 above)
Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: How would you refer the `innerDisposable` otherwise?

Comment (Shachlan) [reply ReactiveObjC/RACSignal+Operations.m]: That wouldn't change.

See [this commit](https://github.com/Lightricks/ReactiveObjC/commit/6945e8210d4044a1e4d246e227c9acc2ad086544) for what I mean.

File: ReactiveObjC/RACSignal+Operations.m (hunk 5 above)
Comment (Shachlan) [reply ReactiveObjC/RACSignal+Operations.m]: @barakwei if you call `subscribeNext:` with the lock taken, and the call to `subscribeNext:` causes values to be sent synchronously, then the lock isn't released before values are sent.

I ran tests with your suggestion, and it seems that the RACCommandSpec gets stuck (probably deadlocked) on several tests.

Comment (Shachlan) [reply ReactiveObjC/RACSignal+Operations.m]: @barakwei if you call subscribeNext: with the lock taken, and the call to subscribeNext: causes values to be sent synchronously, then the lock isn't released before values are sent.

I ran tests with your suggestion, and it seems that the RACCommandSpec gets stuck (probably deadlocked) on several tests.

File: ReactiveObjC/RACSignal+Operations.m (hunk 4 above)
Comment (barakyoresh) [reply ReactiveObjC/RACSignal+Operations.m]: So whichever you prefer. If Schahar wants it asap the counters will also suffice.

Comment (Shachlan) [reply ReactiveObjC/RACSignal+Operations.m]: The time pressure is off - we've circumvented the issue,  so you can finish the PR properly.

Comment (Shachlan) [reply ReactiveObjC/RACSignal+Operations.m]: Also, kudos on the proper Deutsche spelling of my name.

File: ReactiveObjC/RACSignal+Operations.m (hunk 5 above)
Comment (barakwei) [reply ReactiveObjC/RACSignal+Operations.m]: My bad, I thought it was `createSignal:` and not `subscribeNext:`.

**Comment by StatusReport** [comment]
@shachlan it’s on me. Will hopefully get to it this week.

File: ReactiveObjCTests/RACSignalSpec.m (hunk 3 above)
Comment (StatusReport) [reply ReactiveObjCTests/RACSignalSpec.m]: done below

**Comment by StatusReport** [comment]
Gents, PTAL. The locking is not the most elegant thing I've ever written, so if you have an idea on how to make this look better PLMK.

**Comment by Shachlan** [comment]
LGTM

**Comment by StatusReport** [comment]
@alexger PTAL.

**Comment by StatusReport** [comment]
@alexger will you have time for it today?

**Comment by Shachlan** [comment]
ping @alexger

**Comment by alexger** [comment]
I looked at it and had a chat with @StatusReport .
There is a code path where `sendComplete` can be called twice, but I think this is then serialized and handled by the subscriber, so effectively it shouldn't matter.
Another thing is that the weak signals can, in theory, become nil at any time, so the  `if (currentInnerSignal != innerSignal)` check might drop an otherwise legitimate completion.

**Comment by StatusReport** [comment]
Concurrency is **hard**!

Updates:
1. Bug fix: the inner signal could send values after a new inner signal arrived since the old subscription was disposed after the subscription to the new inner signal (and not before).
2. Less potential double completions / errors although the subscribers take care of that.
3. Use spinlocks instead of `NSLock` since the code running while the lock is taken is well defined and not controlled by the user.
4. Locking and unlocking is done linearly and not inside conditions, which should make the code easier to understand.

PTAL again :)

File: ReactiveObjC/RACSignal+Operations.m [hunk 6]
```diff
@@ -852,22 +853,70 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
...
-    RACMulticastConnection *connection = [self publish];
+    __block BOOL outerCompleted = NO;
+    __block BOOL innerCompleted = NO;
+    __block NSUInteger currentInnerIndex = 0;
+
+    RACSerialDisposable *innerDisposable = [[RACSerialDisposable alloc] init];
+    __block os_unfair_lock lock = OS_UNFAIR_LOCK_INIT;
+
+    RACDisposable *outerDisposable = [self subscribeNext:^(RACSignal *signal) {
+      NSCAssert(!signal || [signal isKindOfClass:RACSignal.class], @"-switchToLatest expects signals as values, instead we got: %@", signal);
```
[the remark this answers is not in the document]
Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: Bikeshedding :)

And this project doesn't have 100 char limit. Not sure what to do with new code.

File: ReactiveObjC/RACSignal+Operations.m (hunk 1 above)
Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: As @alexger mentioned, for performance you'd want to avoid `-isKindOfClass:` if `signal` is `nil`.

**Comment by StatusReport** [comment]
@barakwei, @barakyoresh, @alexger - comments?

**Comment by StatusReport** [comment]
Ping.

**Comment by Shachlan** [comment]
ping.

File: ReactiveObjC/RACSignal+Operations.m [hunk 7]
```diff
@@ -852,22 +853,70 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
...
+      signal = signal ?: [RACSignal empty];
+
+      // Ensure no further data is sent from the previous inner signal.
+      [innerDisposable.disposable dispose];
+
+      os_unfair_lock_lock(&lock);
+      NSUInteger innerIndex = ++currentInnerIndex;
+      // In most cases this should be set back to NO, but in case where a new inner signal has
```
Comment (barakyoresh) [inline ReactiveObjC/RACSignal+Operations.m]: Not sure I get the scenario you're trying to protect yourself from. What I think you imagine is
inner completes
inner sets completion flags
outer sends next
outer completes
outer sets completion flags (before handling the next)
outer starts handling next
avoid subscribing to it since outer is already on the track to send completion

If the next is handled before the completion, then the new signal will be subscribed to and the outer won't send completed on completion.

So basically it seels like part of the 

> // Avoid subscribing to a new inner signal which may send more values if we're already going
> // to send complete via the inner or the outer signal.

portion rather than a fear of missing a subscription. So am I missing another scenario this protects against?

File: ReactiveObjC/RACSignal+Operations.m [hunk 8]
```diff
@@ -852,22 +853,70 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
...
+      // completed with the outer, both of the flags should not be reset otherwise we may miss
+      // sending a completion.
+      innerCompleted = innerCompleted && outerCompleted;
```
[the remark this answers is not in the document]
Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: You are right sir. Changed to `alreadyCompleted` and modified the docs.

File: ReactiveObjC/RACSignal+Operations.m [hunk 9]
```diff
@@ -852,22 +853,70 @@ - (RACSignal *)takeUntilReplacement:(RACSignal *)replacement {
...
+      innerDisposable.disposable = [signal subscribeNext:^(id x) {
+        [subscriber sendNext:x];
+      } error:^(NSError *error) {
+        [subscriber sendError:error];
+      } completed:^{
+        os_unfair_lock_lock(&lock);
+        if (innerIndex == currentInnerIndex) {
+          innerCompleted = YES;
+        }
+        NSUInteger sendCompleted = innerCompleted && outerCompleted;
```
[the remark this answers is not in the document]
Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: Indeed.

File: ReactiveObjC/RACSignal+Operations.m (hunk 7 above)
Comment (StatusReport) [reply ReactiveObjC/RACSignal+Operations.m]: Yes, this is the same scenario. I removed the docs here since the variable name is now way better.

**Comment by StatusReport** [comment]
Done. Guys, please review - I want to merge this by end of week.

**Comment by StatusReport** [comment]
🤞
